state/zookeeper/zk.go

```diff
@@ -0,0 +1,328 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package zookeeper
+
+import (
+	"errors"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/hashicorp/go-multierror"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+const defaultMaxBufferSize = 1024 * 1024
+const defaultMaxConnBufferSize = 1024 * 1024
+
+const anyVersion = -1
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+var errMissingServers = errors.New("Servers are required")
+var errInvalidSessionTimeout = errors.New("SessionTimeout is invalid")
+
+type properties struct {
+	Servers           string `json:"servers"`
+	SessionTimeout    string `json:"sessionTimeout"`
+	MaxBufferSize     int    `json:"maxBufferSize"`
+	MaxConnBufferSize int    `json:"maxConnBufferSize"`
+	KeyPrefixPath     string `json:"keyPrefixPath"`
+}
+
+type config struct {
+	servers           []string
+	sessionTimeout    time.Duration
+	maxBufferSize     int
+	maxConnBufferSize int
+	keyPrefixPath     string
+}
+
+func newConfig(metadata map[string]string) (c *config, err error) {
+	var buf []byte
+
+	if buf, err = json.Marshal(metadata); err != nil {
+		return
+	}
+
+	var props properties
+	if err = json.Unmarshal(buf, &props); err != nil {
+		return
+	}
+
+	return props.parse()
+}
+
+func (props *properties) parse() (*config, error) {
+	if len(props.Servers) == 0 {
+		return nil, errMissingServers
+	}
+
+	sessionTimeout, err := time.ParseDuration(props.SessionTimeout)
+	if err != nil {
+		return nil, errInvalidSessionTimeout
+	}
+
+	maxBufferSize := defaultMaxBufferSize
+	if props.MaxBufferSize > 0 {
+		maxBufferSize = props.MaxBufferSize
+	}
+
+	maxConnBufferSize := defaultMaxConnBufferSize
+	if props.MaxConnBufferSize > 0 {
+		maxConnBufferSize = props.MaxConnBufferSize
+	}
+
+	return &config{
+		servers:           strings.Split(props.Servers, ","),
+		sessionTimeout:    sessionTimeout,
+		maxBufferSize:     maxBufferSize,
+		maxConnBufferSize: maxConnBufferSize,
+		keyPrefixPath:     props.KeyPrefixPath,
+	}, nil
+}
+
+type Conn interface {
```

Comment:
What's the purpose of this interface? from what I see Conn is used from the `zk` package.

Reply:
The `Conn` interface was used to generate `MockConn` for testing

Reply:
So can this interface move to `zk_test.go`?

Reply:
The problem is [zk.Conn](https://godoc.org/github.com/samuel/go-zookeeper/zk#Conn) is a structure which hard to be test without a real zookeeper instance. That's why I extract a `Conn` interface for testing.